servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/AbstractTimeoutHttpFilter.java

```diff
@@ -0,0 +1,128 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.http.utils;
+
+import io.servicetalk.concurrent.api.Executor;
+import io.servicetalk.concurrent.api.Single;
+import io.servicetalk.http.api.HttpExecutionStrategy;
+import io.servicetalk.http.api.HttpExecutionStrategyInfluencer;
+import io.servicetalk.http.api.HttpRequestMetaData;
+import io.servicetalk.http.api.StreamingHttpRequest;
+import io.servicetalk.http.api.StreamingHttpResponse;
+
+import java.time.Duration;
+import java.util.concurrent.TimeoutException;
+import java.util.function.Function;
+import javax.annotation.Nullable;
+
+import static io.servicetalk.concurrent.api.Publisher.defer;
+import static io.servicetalk.concurrent.api.Publisher.failed;
+import static io.servicetalk.utils.internal.DurationUtils.ensurePositive;
+import static io.servicetalk.utils.internal.DurationUtils.isPositive;
+import static java.time.Duration.ofNanos;
+import static java.util.Objects.requireNonNull;
+
+abstract class AbstractTimeoutHttpFilter implements HttpExecutionStrategyInfluencer {
+    /**
+     * Establishes the timeout for a given request.
+     */
+    private final TimeoutFromRequest timeoutForRequest;
+
+    /**
+     * If {@code true} then timeout is for full request/response transaction otherwise only the response metadata must
+     * complete before the timeout.
+     */
+    private final boolean fullRequestResponse;
+
+    @Nullable
+    private final Executor timeoutExecutor;
+
+    AbstractTimeoutHttpFilter(final TimeoutFromRequest timeoutForRequest, final boolean fullRequestResponse) {
+        this.timeoutForRequest = requireNonNull(timeoutForRequest, "timeoutForRequest");
+        this.fullRequestResponse = fullRequestResponse;
+        this.timeoutExecutor = null;
+    }
+
+    AbstractTimeoutHttpFilter(final TimeoutFromRequest timeoutForRequest, final boolean fullRequestResponse,
+                              final Executor timeoutExecutor) {
+        this.timeoutForRequest = requireNonNull(timeoutForRequest, "timeoutForRequest");
+        this.fullRequestResponse = fullRequestResponse;
+        this.timeoutExecutor = requireNonNull(timeoutExecutor, "timeoutExecutor");
+    }
+
+    @Override
+    public final HttpExecutionStrategy influenceStrategy(final HttpExecutionStrategy strategy) {
+        return timeoutForRequest.influenceStrategy(strategy);
+    }
+
+    final Single<StreamingHttpResponse> withTimeout(final StreamingHttpRequest request,
+                final Function<StreamingHttpRequest, Single<StreamingHttpResponse>> responseFunction) {
+
+        return Single.defer(() -> {
+            final Duration timeout = timeoutForRequest.apply(request);
+            if (null != timeout && !isPositive(timeout)) {
```

Comment:
do we need todo the pre-checks? ideally the timeout operators handle this case and we can simplify here. we would pay a bit more cost to construct the operators, but if the operators don't handle this everywhere we apply a timeout would have todo a pre-check (which is not ideal).

Reply:
currently - yes. it's a feature of `timeoutForRequest`. if it returns `null`, it means no timeout. If it returns `ZERO` or negative => users want timeout right away.
I will keep it as-is for now. If we will move this logic to timeout operators in a follow-up, we can simplify here. For now, limiting the scope to the filters only

Reply:
The operators do handle it, they explicitly accept negative duration which is expected to occur as a result of calculated timeouts. The appearance here of pre-checks was as a result of earlier review feedback to simplify the flow.

Reply:
`null` is is intended to impact control flow (e.g. `null` means "no timeout") so it is expected to check for `null`. however if there is a timeout value the timeout operators should be able to handle it.

Reply:
I would consider that as a follow-up only if the operator improved. Right now operator will always schedule a task, which creates more overhead compare to this check.

Reply:
followup sounds good: https://github.com/apple/servicetalk/pull/1494